### PR TITLE
Rover: fix motor limits

### PR DESCRIPTION
**Pull request by rmackay9**
This improves Rover/Boat's motor library's setting of the throttle and steering limits flags while reduces I-term buildup and leads to better steering and speed control when the vehicle cannot achieve the desired steering turn rate or speed.  There are two fixes:

1. slew_limit_throttle() was setting the throttle limits but these were always being overwritten by the set_limits_from_input() call made later in output_regular(), output_skid_steering() and output_omni().  This could lead to I-term buildup in the speed controller.  Below is a before-vs-after screenshot from SITL with the MOT_SLEWRATE = 1 (meaning it takes 100 seconds for the throttle to reach full power).  In "before" we see there is constant oscillation while in "after" the vehicle's speed settles nicely on the desired speed.
throttle-lim-before-vs-after
  The fix is to add a new "clear_limits()" method that is called once on each iteration at the top of ::output() and then set_limits_from_input() is modified so it only ever changes the flags to true (never sets them to false).

    This also fixes similar problems with ackermann steering vehicles (e.g. output_regular()) turning while moving slowly.  In master the steering rate controller I-term builds up which leads to overshoot when the sticks are next centered.  Below are screen shots of a before-vs-after test in SITL where the vehicle was put in ACRO mode, throttle was set to 1550 (very low speed) and 30% steering input was provided.  Given time the "Before" does a better job of getting to the desired turn rate but it then continues turning after the sticks are recentered.  In "After" it does not achieve the desired but it also doesn't overshoot later.
ackermann-steering-slow-turns
    Note that I think this method of setting the steering limits when the vehicle falls below MOT_SPD_SCA_BASE (aka "_speed_scale_base") is probably suboptimal.  Instead we should probably calculate (somewhere) the maximum turn rate we can achieve given the current speed and the vehicle's minimum turn radius and then limit the requested rate sent into AR_AttitudeControl::get_steering_out_rate().  I think the max achievable turn rate (in deg/sec) = 360 * speed / (2 * PI * radius).

2. The second change is to output_skid_steering() so that it sets the steering and throttle limit flags if the outputs become saturated (i.e. the mixer can't achieve both the desired steering and throttle output).  The tricky bit though is that during saturation the throttle and/or steering may be scaled back (not limited).  This means that if the controllers provide a higher output (perhaps due to I-term buildup) they will actually get a bit more response.  Originally I thought that because of this we shouldn't set the limit flags... but now I think we should because otherwise we just end up with a "battle of the I-terms"... basically whichever controller has a bigger I-term will win and the other will have unnecessary I-term buildup.  All comments on this are very welcome!

    Below are before-vs-after screen shots from SITL of a skid-steering vehicle's steering controller output in Acro mode (e.g. vehicle is trying to control both speed and turn rate) in which the pilot has requested 90% throttle and 90% right turn which it cannot achieve.  In the "before" we see the I-term building up while in "after" we do not.  Although not shown in the graphs below, the big improvement for users is the vehicle will return to driving straight as soon as the pilot releases the steering stick (in master it keeps turning for a while because of the I-term buildup).
skid-steering-saturation-before-vs-after

As a side note, with Leonard's recent changes to the AC_PID libraries we don't actually need both to have limits for both directions e.g. throttle_upper and throttle_lower could be combined into a single flag.

### CODE REVIEW

**Comment by IamPete1** [comment]
I do have a PR for this https://github.com/ArduPilot/ardupilot/pull/18612. It does need a rebase. The thing that that has that this does not is the E-stop check handling.

**Comment by rmackay9** [comment]
@IamPete1, thanks for the feedback.  Yes, it looks like you spotted the problem before I did.

I think we should merge this PR though still because it's ready (e.g. not in "Draft") and fully tested with reams of testing evidence.  I've put few comments on one of the related PR though so hopefully we can get at least part of yours merged.

File: libraries/AR_Motors/AP_MotorsUGV.h [hunk 1]
```diff
@@ -176,6 +176,9 @@ class AP_MotorsUGV {
     // slew limit throttle for one iteration
     void slew_limit_throttle(float dt);
 
+    // clear all limit flag
+    void clear_limits() { limit.steer_left = limit.steer_right = limit.throttle_lower = limit.throttle_upper = false; }
```
Comment (IamPete1) [inline libraries/AR_Motors/AP_MotorsUGV.h:180]: I'm not really a fan of these little functions if there only used in one place, just makes it harder to see what is going on IMHO.

**Review by IamPete1 — APPROVED** [review]
Looks OK

File: libraries/AR_Motors/AP_MotorsUGV.h (hunk 1 above)
Comment (rmackay9) [reply libraries/AR_Motors/AP_MotorsUGV.h]: your wish is my command.  removed.

**Pull request merged**